Feature/#727 party matching

BlackYps opened the pull request:
This adds the actual matchmaking algorithm to enable matchmaking for teams with more than two players.

BlackYps committed:
Add algorithm for team matchmaking

BlackYps committed:
Fix tests

BlackYps committed:
Calculate deviation based on searches instead of players

BlackYps committed:
Remove redundant code

BlackYps committed:
Sort imports

BlackYps committed:
Change logger string formatting

Askaholic commented on server/config.py:
```diff
@@ -71,6 +71,12 @@ def __init__(self):
+        self.MAXIMUM_RATING_DEVIATION = 300
+        self.TIME_BONUS = 0.025
+        self.MAXIMUM_TIME_BONUS = 0.4
+        self.NEWBIE_BONUS = 0.2
```
How about some comments here. It's not obvious to me what the difference between "rating imbalance" and "rating deviation" is, nor what the units for the bonuses are (are they multipliers? then maybe call them "multiplier" instead of "bonus" which sounds like an additive thing to me).

BlackYps replied:
The bonuses are indeed additive

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+
+    Overview of the algorithm:
+    1. list all the parties in queue by their average rating.
+    2. Take a party and select the neighboring parties alternating between lower
+    and higher until you have 8 players. If the next party to select would leave
+    you with more than 8 players you skip that party and try the next one.
+    3. you now have a list of parties that will be in your potential game. Distribute
+     them into two teams.
```
As I stated on the issue ticket, this is not guaranteed to work. Simply having parties such that total players sum to 8 does not mean that the parties can be distributed into two groups of 4. Take for instance the party sizes (3, 3, 2). IMO step 2 should be making the teams of size 4 directly.

cleborys replied:
While I see the appeal I think that making teams of size 4 this way would systematically produce imbalanced games - at least if you are making teams by grouping together people that are adjacent in rating. (i.e. if there is 8 people sorted by rating then this algorithm has a chance to distribute them as ABABABAB while making teams of adjacent rating would always look like AAAABBBB).

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+                match = self.make_teams(list(participants))
+                game = self.calculate_game_quality(match)
+                possible_games.add(game)
+            except AssertionError:
```
Use a different error type, base `Exception` is probably fine, or you could make your own custom error type. Using assertions for flow control is a bad idea as it is possible to run python in "optimized" mode (`-O` command line switch I think) which will skip all assertions. I would like to have this option in the future, especially if the optimized mode is expanded to do more stuff.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
```
So what's the time complexity here? Should be `O(n^3)` I think?

BlackYps replied:
I think it's `O(n^2)`. Worst case we compare every game with the whole games list, so there we get n * n. (Maybe it is less than `O(n^2)` because the list will be shorter every time but I am not sure how this translates to O notation.) The player checking is `O(m)` where m is the player number in a game so should translate to `O(1)` for big n, right?

cleborys replied:
I assume step 9 is "repeat step 8 until empty" (instead of repeat steps 1-8 until empty)?
I agree that steps 2-4 run in constant time since if the game size is constant and hence step 5 is `O(n)`.
The list of step 6 is again length `n`, so step 7 is `n log(n)` and step 8 is either `O(n)` or `O(1)` depending on how you do it (it's fast if you keep a hash table of all players that are in games that were already selected). Repeating step 8 then adds a factor of `n`, the list getting smaller doesn't make a difference since `n + (n-8) + (n-16) + ...` still asymptotically grows as `n^2`.

So depending on whether you do the removal in step 8 immediately or "on demand" it's either `O(n^2)` or `O(n log n)`

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        searches = list(searches)
+        searches.sort(key=lambda s: s.average_rating, reverse=True)
```
Just use `sorted` if you want to make a new list.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        searches = list(searches)
+        searches.sort(key=lambda s: s.average_rating, reverse=True)
+        self._logger.debug("=== starting matching algorithm ===")
```
Some blank lines in this function would be nice to help readability. These log statements kindof blend in as it is.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+        out_of_bounds_counter = 0
+        while number_of_players < self.team_size * 2 and out_of_bounds_counter < 2:
+            try:
+                candidate = searches[index]
```
First candidate will be the target search itself. Was a little confused at first because to me "neighboring" doesn't sound like it includes yourself. Maybe add a comment here, or a docstring to the function describing this behavior.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+            except IndexError:
+                out_of_bounds_counter += 1
+            i += 1
+            index += i * pow(-1, i)
```
I'm not sure that this function does what you want. As I understand, it's supposed to check searches near a given search, altering between searches that appear earlier in the list and lower in the list, and if not enough searches exist, terminate. However, in python, negative indices are valid and give you items starting from the end of the list. So I believe that as written, this algorithm will treat the list as though it wraps from front to back (but not back to front).

For example if `index` starts at 0. Then on the first few iterations it would look like this:
1. `index = 0`, `i = 0`
2. `index = -1`, `i = 1` (!`index -1` references the last element in the list! i.e. the lowest rated search)
3. `index = 1`, `i = 2`
4. `index = -2`, `i = 3` (! now index references the second lowest rated search)

I'd suggest using iterators instead. Generally speaking, iterating through lists using indices is "un-pythonic". You could do something like this instead:
```python3
higher = searches[:index:-1]
lower = searches[index:]
i = 0
while condition:
    candidate = next(higher if i % 2 else lower, None)
    if candidate is None:
        # Handle exhausted list
    i += 1
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+        return participants
+
+    def make_teams(self, searches: List[Search]) -> Tuple[Search, Search]:
+        avg = CombinedSearch(*searches).average_rating
+        team_target_strength = CombinedSearch(*searches).cumulated_rating / 2
```
Not a fan of using the search objects in this way, just compute the values directly with generator expressions:
```suggestion
        avg = statistics.mean(search.avg_rating for search in searches)
        team_target_strength = sum(search.cumulated_rating for search in searches) / 2
```

BlackYps replied:
This won't work for the average rating, because we need to take the different number of players in the searches into account.

Askaholic replied:
Well, write a function that does that then. But creating a new search object just to access one property which will turn into a function call through python magic is not good. You could even have the property call your function instead so the code isn't duplicated.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+            self._logger.debug("building game for %s", repr(search))
+            participants = self._pick_neighboring_players(searches, index)
+            try:
+                match = self.make_teams(list(participants))
```
Copying list for no reason. IMO it should be the callee's responsibility to make a copy if it needs to do that to prevent modifying arguments.
```suggestion
                match = self.make_teams(participants)
```

Askaholic replied:
In this particular case actually, the function *could* take ownership of the participants list and modify it directly, because I don't see it being used anywhere else. Probably better not to do that though.

BlackYps replied:
So if I just delete the copy in the call like you suggested will that make the function take ownership of the list like you said in your last comment, or does it need extra steps to do that?
I.e. should I just delete the copy in the call and leave the function as is, or should I make additional changes in the function?

Askaholic replied:
Well sadly there isn't a concept of ownership built into the language, as everything in python is a shared reference. I mean "Taking ownership" in a conceptual sense. It would just mean you make sure you don't use the argument after you've passed it to the function because you know that the function will assume that it owns that data and modify it. You would have to document the behavior clearly in the docstring.

Something like this for example:
```python3
takes_foo(makes_foo())
```
The `takes_foo` function can do whatever it wants to the output from `makes_foo` because it's never assigned to a variable so it can't even be used in the outer scope.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+            searches.remove(filler)
+        else:
+            team_a, searches = self.run_karmarkar_karp_algorithm(searches)
+        team_b.extend(searches)
```
If you do it like this, then you don't need to call `remove`.
```suggestion
        team_b.extend(search for search in searches if search not in team_a)
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+    def run_karmarkar_karp_algorithm(self, searches):
+        self._logger.debug("Running Karmarkar-Karp to partition the teams")
+        containers = []
+        for s in searches:
```
Use a list comprehension instead.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+            #  elem1 is always bigger than elem2
+            container = Container(elem1.rating - elem2.rating, [elem1, elem2])
+            containers.append(container)
+            containers.sort(key=lambda c: c.rating)
```
You could use a `SortedList` from the `sortedcontainers` package. (need to add it to pipfile)

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+        containers_b = []
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while len(containers_a) > 0 or len(containers_b) > 0:
```
```suggestion
        while containers_a or containers_b:
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while len(containers_a) > 0 or len(containers_b) > 0:
+            for e in containers_a:
```
WTF is this dark voodoo magic?!??! Definitely need some comments here

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,262 @@
+        if not searches_dict[1]:
+            self._logger.warning("given searches are impossible to split in even teams because of party sizes")
+            raise AssertionError
+        searches_dict[1].sort(key=lambda s: s.cumulated_rating, reverse=True)
```
Maybe put in the docstring that this function modifies its arguments

Askaholic replied:
TBH I'm not sure why this function is so complicated. Can't you just iterate through this list and find the best match according to some metric? I feel like this should look very similar to a "max" implementation:
```python3
if not the_list:
    return None

iterator = iter(the_list)
best = next(iterator)
for item in iterator:
    if item > best:
        best = item
return best
```

BlackYps replied:
It's partly because the code looks messier because it's not using an inteator. The other part of the answer is that the condition to abort is not so easy. I can't check if a metric is fulfilled, but I need to check for better solutions until I find a worse one and know because of this what the best solution was.

Askaholic replied:
Why do you need to find a worse one? You only care about the best right?

BlackYps replied:
I need to find the player that gives me the team that is closest to the intended rating. It's basically an optimization problem. The perfect solution (getting exactly the intended average) will generally not happen. So I have to try them one by one as long as the solution gets incrementally better. I only know what the best solution is, when I can be sure that all following candidates will provide a worse solution.

Askaholic reviewed (commented):
I've only made it about 60% of the way through, but hopefully you can make use of these comments everywhere. 

The whole thing with picking 8 players and then trying to split them into teams afterwards when it can possibly fail still seems weird to me because then you are left with searches for which there is no candidate game with them as the target. Wouldn't it make more sense to come up with exactly 1 target game per search, or at least fail the candidate games due to insufficient quality instead of choosing a poor combination of other searches such that the group cant be split into 2 teams?

I think it would be good to test some of your helper/intermediate functions too. You can even use `hypothesis` to do property based testing, which in my experience is pretty good about catching common edge cases (like empty lists as input).

BlackYps commented:
> The whole thing with picking 8 players and then trying to split them into teams afterwards when it can possibly fail still seems weird to me because then you are left with searches for which there is no candidate game with them as the target. Wouldn't it make more sense to come up with exactly 1 target game per search, or at least fail the candidate games due to insufficient quality instead of choosing a poor combination of other searches such that the group cant be split into 2 teams?

I agree that it is a bit weird. Do you have a good solution to this problem? In my tests I saw that it happens seldom and I hope that if it fails the search would still be included in one of the neigboring games.

Askaholic commented:
Dunno if these are *good* solutions, but here are some ideas.

What if you just do the sort by rating thing twice:
1. Sort searches by avg rating, and group searches into teams of 4 people
2. Sort these teams of 4 people by avg rating, and group into games
3. Pick highest quality game

Or maybe check the quality of every neighboring team. So if I number off a bunch of teams by rating rank, `[1st, 2nd, 3rd, 4th]`, you could check the game qualities between `1st - 2nd` then `2nd - 3rd` then `3rd - 4th`, etc.

I would also be curious to see what the advantage is of only picking 1 game each iteration, vs picking as many as you can that are above the quality threshold and don't overlap.

BlackYps commented:
making static teams first leaves out a lot of possibilities. Here are the games generated by test_team_matchmaker_algorithm:

```
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p10', 'p9', 'p8', 'p6'], 1244.75) vs Search(['p7', 'p2', 'p3', 'p4'], 1247) rating disparity: -9 quality: 0.554257
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p5', 'p2', 'p16', 'p15'], 1085.5) vs Search(['p6', 'p12', 'p11', 'p1'], 1084.25) rating disparity: 5 quality: 0.870633
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p5', 'p2', 'p16', 'p15'], 1085.5) vs Search(['p6', 'p12', 'p11', 'p1'], 1084.25) rating disparity: 5 quality: 0.870633
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p12', 'p11', 'p1', 'p4'], 1140.75) vs Search(['p16', 'p15', 'p5', 'p0'], 1138.75) rating disparity: 8 quality: 0.738119
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p5', 'p2', 'p16', 'p15'], 1085.5) vs Search(['p6', 'p12', 'p11', 'p1'], 1084.25) rating disparity: 5 quality: 0.870633
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p10', 'p9', 'p8', 'p3'], 1316.5) vs Search(['p7', 'p2', 'p14', 'p13'], 1248.5) rating disparity: 272 quality: 0.336618
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p14', 'p13', 'p0', 'p1'], 1244) vs Search(['p4', 'p5', 'p12', 'p11'], 1147.25) rating disparity: 387 quality: 0.256705
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p10', 'p9', 'p8', 'p4'], 1301.25) vs Search(['p14', 'p13', 'p3', 'p0'], 1298) rating disparity: 13 quality: 0.883450
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p14', 'p13', 'p0', 'p1'], 1244) vs Search(['p4', 'p5', 'p12', 'p11'], 1147.25) rating disparity: 387 quality: 0.256705
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p10', 'p9', 'p8', 'p3'], 1316.5) vs Search(['p14', 'p13', 'p7', 'p4'], 1306.75) rating disparity: 39 quality: 0.854256
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p10', 'p9', 'p8', 'p5'], 1269) vs Search(['p4', 'p14', 'p13', 'p0'], 1282.75) rating disparity: -55 quality: 0.723366
DEBUG    TeamMatchMaker:team_matchmaker.py:67 game: Search(['p0', 'p16', 'p15', 'p6'], 1114.5) vs Search(['p12', 'p11', 'p1', 'p5'], 1108.5) rating disparity: 24 quality: 0.750042
```
 For good games the rating disparity is extremely low. I doubt that you can get this when you determine the teams earlier. Also you can see that the team composition is different for basically every game.

Maybe we could not only check for not only for having exactly 8 people but filling two teams of 4 and then have the exact team composition sorted out later. Or we manually filter for the "bad" compositions. There is only one for 3v3 and 4v4: [2, 2, 2] and [3, 3, 2] respectively.

Still, what exactly do we do if we detect that? We would need some sort of alternative algorithm for picking the participants if we detected a problem in the first try.

Askaholic commented:
One thing I'm noticing in that output is that it seems to check the same matches a lot.
```
3 times - Search(['p5', 'p2', 'p16', 'p15'], 1085.5) vs Search(['p6', 'p12', 'p11', 'p1'], 1084.25)
2 times - Search(['p14', 'p13', 'p0', 'p1'], 1244) vs Search(['p4', 'p5', 'p12', 'p11'], 1147.25)
```

Askaholic commented:
Also, how does an exhaustive search perform? Like if you were to simply check every possible game, how large does the queue need to get (number of parties) before the performance becomes a bottleneck?

BlackYps commented:
Having the same matches multipe times is expected to a degree. Close to the end of the list of players you will get the same participants multiple times. But I should probably fix the problem with the negative indices first before we dive deeper into that.

Do you mean _every_ possible game? A rough estimate: let's say we have 20 single player searches. There are 125970 possibilities to choose 8 of these 20. Then there are 70 possible combinations to split 8 players into two teams of four. So we get a little under 9 million games. We need to calculate the quality for all of them. I just tested it and it took 1.8 seconds for 9000 games, so it would take 30 minutes for 9 million.

BlackYps committed:
Use custom exception instead of AssertionError

BlackYps committed:
Improve documentation

Askaholic commented:
Yea that's too much :P

BlackYps committed:
Refactor participant picking

BlackYps committed:
Add small code improvements

BlackYps committed:
Refactor _find_most_balanced_filler

BlackYps committed:
Fix imports

BlackYps committed:
Use sortedcontainers package

Askaholic commented on server/config.py:
```diff
@@ -72,6 +72,14 @@ def __init__(self):
 
+        # Values for the custom (i.e. not trueskill) game quality metric used by the matchmaker
+        self.MINIMUM_GAME_QUALITY = 0.5
+        self.MAXIMUM_RATING_IMBALANCE = 600  # Difference of cumulated rating of the teams
```
Just put the comments on the line above the thing they're describing

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+                               repr(game.match[0]),
+                               repr(game.match[1]),
+                               game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                               game.quality)
```
Fix formatting. The args should be on the next line (see contributing guidelines)

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+
+        failed_last_time = False
+        while number_of_players < self.team_size * 2:
+            candidate = next(lower if i % 2 else higher, None)
```
You could do something like this:
```suggestion
            candidate, prev = next(lower if i % 2 else higher, None), candidate
```
Then your break condition might be simpler:
```python3
if candidate is None:
    if prev is None:
        raise NotEnoughPlayersException
    continue
```

Askaholic replied:
Could also use a boolean toggle instead of a counter
```suggestion
            candidate, select = next(lower if select else higher, None), not select
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+        participants = []
+        i = 0
+        number_of_players = 0
+        lower = searches[:index]
+        lower = iter(lower[::-1])
```
I think it's fine to combine here

BlackYps replied:
Combining actually alters the behaviour. I added a comment to explain it.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+        Attempts to partition the given searches into two teams of the appropriate team size
+        while also trying that both teams have the same cumulated rating.
+        Raises UnevenTeamsException if one of the teams doesn't have the right size.
+        :param: The searches to partition. The function will alter this list!
+        :return: The two teams
```
Use markdown instead of rst. See `GameConnection` for examples. This is because our auto-documentation expects markdown

BlackYps replied:
Maybe I am blind but in `GameConnection` the comments seem to have the exact same syntax?

Askaholic replied:
The base branch is probably out of date. Just look on `develop`

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+
+        self._logger.debug("participating searches by player size:")
+        for i in range(self.team_size, 0, -1):
+            self._logger.debug("%i players: %s", i, searches_by_size[i])
```
I'm not a fan of these debug loops since they won't become nops in debug mode. It would be better to have a single log statement that prints out the whole datastructure. If you want to keep this for your own testing, fine but please change it before we merge.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+            new_avg_delta = abs(avg - new_team_avg)
+            self._logger.debug("delta with %s is %s (avg is %s)", [candidate], avg_delta, team_avg)
+            self._logger.debug("new delta with %s is %s (avg is %s)", [item], new_avg_delta, new_team_avg)
+            if new_avg_delta > avg_delta:
```
Ok, this is what I thought last time. This function just finds the candidate with the minimum average rating difference. I don't think there's really any need to presort the list here, just iterate through it once and then return the candidate with the lowest rating difference. 

Presorting just turns this from a `O(n)` operation into an `O(n * log n)` operation. The only reason I could see for doing this would be if computing the `delta` metric is somehow way more expensive than computing the sort key. But I don't think that's the case here, I think it's actually just doing a bunch of unnecessary work. Do what I suggested last time and the function will run faster and require less code.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+        self._logger.debug("used %s as filler", [candidate])
+        return candidate
+
+    def calculate_game_quality(self, match: Match) -> Game:
```
Something is misnamed here. The function suggest it returns a quality but instead it returns a `Game` which is essentially just a tuple of match and quality (except not actually because it's a class instead).

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+        games = SortedList(games, key=lambda gme: gme.quality)
+
+        matches: List[Match] = []
+        while len(games) > 0:
```
Embrace the Python way :P
```suggestion
        while games:
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+    def _pick_best_noncolliding_games(self, games: List[Game]) -> List[Match]:
+        for game in list(games):
+            if game.quality < config.MINIMUM_GAME_QUALITY:
+                games.remove(game)
```
Since you're copying `games` anyways, you might as well do this in a way that doesn't modify the original. Plus I believe this way is faster.
```suggestion
        games = [game for game in games if game.quality >= config.MINIMUM_GAME_QUALITY]
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+                games.remove(game)
+        self._logger.debug("%i games left after removal of games with quality < %s",
+                           len(games), config.MINIMUM_GAME_QUALITY)
+        games = SortedList(games, key=lambda gme: gme.quality)
```
```suggestion
        games = SortedList(games, key=lambda game: game.quality)
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+                           len(games), config.MINIMUM_GAME_QUALITY)
+        games = SortedList(games, key=lambda gme: gme.quality)
+
+        matches: List[Match] = []
```
Not sure about here, but usually mypy is OK at figuring out the types of variables like this.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+            used_players = set()
+            for search in g.match:
+                for player in search.players:
+                    used_players.add(player)
```
This can become a comprehension (technically it's a function call with a generator expression). I think something like this is the correct order for nested loops:
```suggestion
            used_players = set(player for search in g.match for player in search.players)
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+                for player in search.players:
+                    used_players.add(player)
+            self._logger.debug("used players: %s", [p.login for p in used_players])
+            for game in list(games):
+                for search in game.match:
+                    if not set(search.players).isdisjoint(used_players):
+                        games.remove(game)
+                        self._logger.debug("removed game: %s", game.match)
+                        break
```
Turns out `isdisjoint` actually works on any iterable, so there's no need to construct a new set each time:
```suggestion
            games = [
                game for game in games
                if used_players.isdisjoint(
                    player for search in game.match for player in search.players
                )
            ]
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,292 @@
+            newbie_bonus + time_bonus, rating_disparity, fairness, deviation, uniformity, quality)
+        return Game(match, quality)
+
+    def _pick_best_noncolliding_games(self, games: List[Game]) -> List[Match]:
```
I don't think this function quite does what it claims because it's using a greedy approach. The first game that meets the quality threshold will always be added to the matches even if that results in a suboptimal list of matches. For instance suppose your games are:

- game 1: `[p1, p2]`
- game 2: `[p1, p3]`
- game 3: `[p2, p4]`

And all qualities are equal. 

The function will return `[game 1]` because it appears first in the list, however, the best result would be to return `[game 2, game 3]` since they don't conflict with eachother and this result yields more games.

Obviously this is a difficult optimization problem in and of itself (I believe it is equivalent to the knapsack problem which we know is NP hard, i.e. cannot be done in polynomial time). Maybe we don't need to find the best list of matches but only a list that is good enough, or not the worst, but I just wanted to point that out since the function name claims that it finds the best.

BlackYps replied:
Depends on your definition of "best" I guess :P This one ensures that the best quality game is always selected.

I actually wanted to add more code later to find a better solution to this optimization problem that yields more games. For now I stuck to the greedy approach to just have something so I can test the rest.

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -16,6 +17,28 @@
 OnMatchedCallback = Callable[["Search", "Search"], Any]
 
 
+class Game:
```
Not a fan of this name because we already have a class called `Game` in the `games` module.

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -16,6 +17,28 @@
+    """
+    Represents a potential game
+    """
+    def __init__(
+        self,
+        match: Match,
+        match_quality: float
+    ):
```
I would use a `typing.NamedTuple` as long as having immutable objects is OK, and if not then maybe you want a `dataclass`.

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -16,6 +17,28 @@
+
+
+def get_average_rating(searches):
+    ratings = []
+    for s in searches:
+        for mean, dev in s.raw_ratings:
+            rating = mean - 3 * dev
+            ratings.append(rating)
+    return statistics.mean(ratings)
```
This could be a one-liner.
```suggestion
    return statistics.mean(mean - 3 * dev for s in searches for mean, dev in s.raw_ratings)
```

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -88,6 +111,24 @@ def ratings(self):
+            mean, dev = player.ratings[self.rating_type]
+            rating = mean - 3 * dev
+            cumulated_rating += rating
+        return cumulated_rating
```
Again this could be a one liner. Use `sum` and a generator expression

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -88,6 +111,24 @@ def ratings(self):
+            mean, dev = player.ratings[self.rating_type]
+            rating = mean - 3 * dev
+            ratings.append(rating)
+        return statistics.mean(ratings)
```
again, one-liner

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -113,7 +154,7 @@ def boundary_75(self):
     @property
     def failed_matching_attempts(self) -> int:
-        return self._failed_matching_attempts
+        return self._failed_matching_attempts * len(self.players)
```
This seems very strange to me. I don't see why we would be prioritizing larger searches who have been in the queue the same amount of time as smaller searches. If some special adjustment for larger searches needs to happen then IMO it belongs in algorithm and not here.

BlackYps replied:
We are not prioritizing larger searches. I am looking at cumulated time spent in queue. At the moment four guys queueing single aquire the time bonus four times faster than if they queue up as four premades, because the bonus is only given per search. I want the time bonus per player to be always equal.

Askaholic replied:
Can you add the player count multiplier in the code that calls this. Or at the very least rename this property because IMO this behaviour is confusing:

```python3
s1 = Search([p1])
s1.register_failed_matching_attempt()
print(s1.failed_matching_attempts)
# 1

s2 = Search([p1, p2, p3, p4])
s2.register_failed_matching_attempt()
print(s2.failed_matching_attempts)
# 4

BlackYps replied:
yeah, I just realized it makes way more sense to add the multiplier when I calculate the quality

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -273,13 +314,21 @@ def players(self) -> List[Player]:
     @property
     def failed_matching_attempts(self) -> int:
-        return max(search.failed_matching_attempts for search in self.searches)
+        return sum(search.failed_matching_attempts for search in self.searches)
```
Again, seems weird to me.

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+
+
+def make_searches(ratings, player_factory):
+    searches = []
+    for i, r in enumerate(ratings):
+        searches.append(Search([player_factory(r + 300, 100, name=f"p{i}")]))
```
Pretty much everywhere where you have this pattern it can be replaced by a comprehension/generator expression

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+    return searches
+
+
+def test_team_matchmaker_performance(player_factory, bench, caplog):
```
Not sure how good these performance tests are. Maybe we should put them behind a marker so they aren't run by default. I've had issues with the one I wrote being flaky in the CI builds.

BlackYps replied:
Sounds good. how would I do that?

Askaholic replied:
Probably need to define a marker in the main `conftest` file (just copy how the other ones are done). Then you just add the decorator:

```python3
@pytest.mark.my_marker
def test_whatever():
    ...
```

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+    searches = [Search([player_factory(1500, 500, ladder_games=1)]) for _ in range(num_searches)]
+
+    with bench:
+        TeamMatchMaker(4).find(searches)
```
This is also including the time it takes to construct the `TeamMatchMaker` object (not that it will be relevant if the number of searches is large enough)

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+
+    assert len(matches) == 2
+    for match in matches:
+        assert TeamMatchMaker(4).calculate_game_quality(match).quality > config.MINIMUM_GAME_QUALITY
```
Maybe reuse the same `TeamMatchMaker` object

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+
+    matches = TeamMatchMaker(4).find(s)
+
+    assert len(matches) == 2
+    for match in matches:
```
Assuming the algorithm is deterministic, it would probably be be better to `assert matches == [...]`. I guess it depends on how annoying it is to assign the searches to variables and stuff.

BlackYps replied:
I wanted to have the test more generic, because we don't care so much about a specific result, but about a good result. But maybe the better approach is to have the unit test very narrow and test the more general functionality with Hypothesis tests.

Askaholic replied:
Yea if you want to test properties then hypothesis is great. For me tests that check a specific example are also useful for documentation purposes, and it's a lot easier to read when you assert equality.

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker(4).calculate_game_quality((team_a, team_b))
+
+    assert game.quality == 0.0
```
Where are the tests for the private methods!? You know, like the one that had a bug because it was using negative indices. That could have been found with a unit test or two.

BlackYps replied:
imfine

BlackYps replied:
I'll add some

BlackYps committed:
Refactoring

BlackYps committed:
Improve code style

BlackYps committed:
Reuse teammatchmaker object in tests

BlackYps committed:
Use namedTuple instead of a class for game

BlackYps committed:
Add more tests

BlackYps committed:
Also return unmatched searches

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,286 @@
+from ..search import CombinedSearch, Match, Search, get_average_rating
+from .matchmaker import Matchmaker
+
+Game = namedtuple('Game', ('match', 'quality'))
```
I prefer `typing.NamedTuple`. There should be some examples in various `typedef.py` files.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,286 @@
+                self._logger.warning("Failed to assign even teams. Skipping this game...")
+
+        self._logger.debug(
+            "got %i games\n" % len(possible_games) + "\n".join(
```
Use f-strings instead of `%` operator.:
```suggestion
            f"got {len(possible_games)} games\n\n".join(
```

BlackYps replied:
I thought you dislike the f-strings because they are less performant? Do you want me to change only this one specifically or all?

Askaholic replied:
Well, in general I am a fan of f-strings, but the logger library doesn't support that style of formatting in log messages afaik. Which means if you use f-strings in log calls the string formatting will happen eagerly even if the string is never printed. This means ever so slightly worse performance that you will never notice except possibly in very tight loops. But I still prefer the '%' way for "correctness" and consistency because we do it in all log statements in this project.

However, in this case you are eagerly formatting the string anyways because you want to print out many lines of output, so I'd just go ahead and use an f-string.

Now that I think of it though. There might be an API for checking what the log level of the logger is so you could do that and then put your for loop that prints a bunch of debug inside of an if statement.

```python3
if self._logger.isEnabledFor(logging.DEBUG):
    # Put as many debug printing loops as you want here
    # because they won't run in INFO logging mode
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,286 @@
+            i += 1
+            if candidate is None:
+                if prev is None:
+                    raise NotEnoughPlayersException
```
I think you're supposed to raise objects, not classes.
```suggestion
                    raise NotEnoughPlayersException()
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,286 @@
+        lower = iter(searches[index - 1::-1])
+        higher = iter(searches[index:])
+        i = 0
+        candidate = next(higher)
```
This can raise `StopIteration` if higher is empty. You can use the second argument to return a default instead.
```suggestion
        candidate = next(higher, None)
```

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,248 @@
+    return make
+
+
+matchmaker = TeamMatchMaker(4)
```
Well you don't want to make it a global variable because that could cause cross-test interference. I would make one per test, or you could make it a fixture in which case pytest will take care of making 1 per test for you.

BlackYps committed:
Change bonus calculation

BlackYps committed:
Code style improvements

BlackYps committed:
Add test for game picking

BlackYps committed:
Use boolean in pick_neighboring_players

BlackYps committed:
Fix test for failed matching attempts

BlackYps commented:
> I would also be curious to see what the advantage is of only picking 1 game each iteration, vs picking as many as you can that are above the quality threshold and don't overlap.

Interestingly it just increases the average wait time by a little, but it has no positive effect on the game quality. I guess this is because the advantage that we don't launch a second game with lower quality is counteracted by the algorithm being more lenient with the quality if it can match people with a higher wait time.
> One thing I'm noticing in that output is that it seems to check the same matches a lot.

Now that the participant picking works without bugs it is evident that the duplicate matches are at the end of the search, which makes sense. I decided to not try to filter those out, because it would be more complicated and also slower than to just swat them in the end, when we check for noncolliding games.

BlackYps committed:
Refactor debug messages

BlackYps commented:
I noticed that we probably want to have different configuration values for the different queues. It is for example much harder to tightly balance a 2v2 than a 4v4. How could we do that?

BlackYps committed:
Accomodate for changes by rebase

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        possible_games = []
+        for index, search in enumerate(searches):
+
+            self._logger.debug("building game for %s", repr(search))
```
BTW I think you can do `%r` and it will use `repr` instead of `str`

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        higher = iter(searches[index:])
+        pick_lower = True
+        candidate = next(higher, None)
+        participants = [candidate]
```
It's possible for `candidate` to be `None` here if `index >= len(searches)`, although this will never be called with such an index. I think I would still add an `assert` statement or two at the top of the function to put these preconditions into code. Maybe even `assert candidate is not None` because I suspect MyPy will complain otherwise.

cleborys replied:
Defining `candidate` as `searches[index]` (and adjusting `higher` accordingly) might also make it more clear what is happening.

BlackYps replied:
I think Cleborys' solution should take care of all these points

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+    def make_teams(self, searches: List[Search], team_size: int) -> Tuple[Search, Search]:
+        """
+        Attempts to partition the given searches into two teams of the appropriate team size
+        while also trying that both teams have the same cumulated rating.
```
"cumulative" instead of "cumulated"?

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            raise UnevenTeamsException()
+
+        avg = get_average_rating(searches)
+        team_target_strength = sum(search.cumulated_rating for search in searches) / 2
```
Same here. I don't think "cumulated" is a word.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                           team_a, combined_team_a.cumulated_rating, combined_team_a.average_rating)
+        self._logger.debug("team b: %s cumulated rating: %s average rating: %s",
+                           team_b, combined_team_b.cumulated_rating, combined_team_b.average_rating)
+        if not len(combined_team_a.players) == team_size:
+            raise UnevenTeamsException()
+        if not len(combined_team_b.players) == team_size:
+            raise UnevenTeamsException()
```
These checks could happen before making the CombinedSearch objects. Only thing is you wouldn't get the debug logs for uneven teams, not sure how important that is.

BlackYps replied:
I'd rather leave them where they are, so you can use the logs to better understand what went wrong, should that be necessary.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        elem1 = containers.pop()
+        elem2 = containers.pop()
+        while True:
+            #  elem1 is always bigger than elem2
```
```suggestion
            # elem1 is always bigger than elem2
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        # Karmarkar-Karp works only for positive integers. By adding 5000 to the rating of each player
+        # we also strongly incentivise the algorithm to give both teams the same number of players
+        containers = SortedList([Container(5000 * len(s.players) + s.cumulated_rating, [s]) for s in searches],
+                                key=lambda c: c.rating)
```
```suggestion
        containers = SortedList(
            [Container(5000 * len(s.players) + s.cumulated_rating, [s]) for s in searches],
            key=lambda c: c.rating
        )
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        # different teams (The higher one into the team the container is in).
+        team_a = []
+        team_b = []
+        containers_a = []
+        containers_b = []
```
Have you checked what the performance impact of making these be `set`s or `deque`s is? I think they will have only a few elements in them right? So maybe it's negligible or even lists are faster.

cleborys replied:
It's at most 2 * team_size elements I guess

BlackYps replied:
That's correct. I don't think there is a worthwhile performance impact with these small lengths.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            max_size = max(searches_by_size.keys())
+            self._logger.debug("participating searches by player size:")
+            for i in range(1, max_size + 1):
+                self._logger.debug("%i players: %s" % (i, searches_by_size[i]))
```
```suggestion
                self._logger.debug("%i players: %s", i, searches_by_size[i])
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            raise UnevenTeamsException()
+
+        iterator = iter(searches_dict[1])
+        candidate = next(iterator)
+        old_team_avg = get_average_rating([search, candidate])
+        old_avg_delta = abs(avg - old_team_avg)
+        self._logger.debug("delta with %s is %s (avg is %s)", [candidate], old_avg_delta, old_team_avg)
```
You could simplify this duplication by just setting `old_avg_delta` to `math.inf` here. I'd probably also rename it to `min_delta` since that's what you're trying to find.

I think this this whole function could also be a one liner (only you will lose the logging):
```python3
return min(
    searches_dict[1], 
    key=lambda item: abs(avg - get_average_rating([search, item]))
)
```

BlackYps replied:
lol, you are right, I didn't even notice I basically reimplemented the min function :D

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        iterations there is no benefit to use a more sophisticated algorithm.
+        """
+        games = [game for game in games if game.quality >= config.MINIMUM_GAME_QUALITY]
+        self._logger.debug("%i games left after removal of games with quality < %s",
+                           len(games), config.MINIMUM_GAME_QUALITY)
```
I really dislike this way of formatting. I much prefer this:
```suggestion
        self._logger.debug(
            "%i games left after removal of games with quality < %s",
             len(games), config.MINIMUM_GAME_QUALITY
         )
```
I don't really mind if you put all args on their own line or have some lines with multiple args as long as it remains readable.

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            self._logger.debug("used players: %s", [p.login for p in used_players])
+            games = [
+                game for game in games
+                if used_players.isdisjoint(
+                    player for search in game.match for player in search.players
+                )
```
Maybe it would be worth refactoring this to a function or property so you could do something like:
```suggestion
                if used_players.isdisjoint(game.all_players)
```

Askaholic commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            if matches:
+                self._logger.debug("Chosen games:")
+            for match in matches:
+                self._logger.debug("%s vs %s " % (repr(match[0]), repr(match[1])))
```
```suggestion
                self._logger.debug("%s vs %s ", repr(match[0]), repr(match[1]))
```

Askaholic commented on server/matchmaker/search.py:
```diff
@@ -308,3 +336,9 @@ def cancel(self):
+        """
+        Returns the searches of which this CombinedSearch is comprised
+        """
+        return list(self.searches)
```
What if you have something nested like `CombinedSearch([CombinedSearch([...], Search())`? Should it flatten the searches out like this? Or just go down one layer?
```suggestion
        return list(s for search in self.searches for s in search.get_original_searches())
```

BlackYps replied:
I only want to go down one layer. I don't know a better wording for the doc string, though.

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,321 @@
+        assert len(teams[0].players) == team_size
+        assert len(teams[1].players) == team_size
+    except UnevenTeamsException:
+        assert True
```
Maybe this would be a good place to use `assume`: https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions
```suggestion
        assume(False)
```
This should (in theory) make hypothesis waste less time trying examples that are not relevant for the test.

BlackYps replied:
This actually increases the time the test uses, because it marks examples as invalid and so it has to generate more examples to find 100 valid ones. I created a new strategy (and assumption) that hopefully generates more suitable data.

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,321 @@
+@given(player=st_players())
+def test_maximum_game_quality_for_even_teams(player):
+    search = Search([player])
+    team_a = CombinedSearch(*[search, search, search, search])
+    team_b = CombinedSearch(*[search, search, search, search])
```
In python you can multiply lists by integers:
```suggestion
    team_a = CombinedSearch(*[search] * 4)
    team_b = CombinedSearch(*[search] * 4)
```

BlackYps replied:
Wait, wat? Python never seizes to amaze...

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,321 @@
+    num_newbies = sum(search.has_newbie() for search in team_a.get_original_searches())
+    num_newbies += sum(search.has_newbie() for search in team_b.get_original_searches())
+
+    assert abs(quality_before + 6 * config.TIME_BONUS + num_newbies * config.NEWBIE_BONUS - quality_after) < 0.0000001
```
What exactly is this testing? Isn't this just duplicating the implementation in the test? Kindof like:

```python3
def add2(a, b):
    return a + b

def test_add2(a, b):
    assert add2(a, b) == a + b

BlackYps replied:
You are not wrong. I added this test to ensure that nobody accidentally breaks the time bonuses.
The assertion is a bit ugly because I get rounding errors.

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,321 @@
+        participants = TeamMatchMaker().pick_neighboring_players(searches, index, 4)
+        assert sum(len(search.players) for search in participants) == 8
+    except NotEnoughPlayersException:
+        assert True
```
```suggestion
        assume(False)
```

Askaholic commented on tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py:
```diff
@@ -0,0 +1,321 @@
+
+@given(st_game_candidates_list())
+def test_pick_noncolliding_games(games):
+    if not games:
+        assert TeamMatchMaker().pick_noncolliding_games(games) == []
+        return
```
Maybe this should be its own test

Askaholic replied:
Then you could just `assume(games)` here.

Askaholic reviewed (commented):
Its looking pretty good!

Askaholic commented:
> I noticed that we probably want to have different configuration values for the different queues. It is for example much harder to tightly balance a 2v2 than a 4v4. How could we do that?

I guess that info would need to go into the database.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+    from the matchmaker. The quality is not the trueskill quality!
+    """
+    match: Match
+    quality: float
```
Could rename it to `custom_quality` or `adjusted_quality` if it's important to keep in mind that this is not the trueskill quality. (I think it's fine as is.)

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+    quality: float
+
+
+class Container:
```
The name "Container" leaves me wondering at what this is, exactly, in particular what `content` will be. Maybe a game candidate? I assume that will become clear when I'll read on but it's nice to choose names that don't leave readers guessing at your intentions.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+
+        self._logger.debug("========= starting matching algorithm =========")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
```
Looking at `search.py` I see that you are using the displayed rating `mean - 3 * dev` here and I think that's a bad idea. I *really, really* think of that strange way to get a rating from the mean and dev as a "hack" so that players see their rating increase over time rather than fluctuate and I don't think it has any significance in trueskill (other than "we are 96% sure that you will perform better than this rating in a given game", or whatever the three confidence intervals certainty is). This feels like intentionally feeding worse data into your algorithm.
I would simply use the average mean, which is trueskill's best guess for your skill.

Of course that doesn't account for new players who start at a (1500, 500) rating that is obviously (and intentionally) too high. To account for that in the matchmaker we have previously treated anyone with ten games or less as if they had a hard-coded mean, you can probably search for "newbie_adjusted_rating" or for the config parameter "NEWBIE_BASE_MEAN".
(I can't recall whether we just used `NEWBIE_BASE_MEAN` directly or interpolated between `NEWBIE_BASE_MEAN` and their current trueskill mean over the course of the first ten games.)

BlackYps replied:
If we wanted to fully utilize trueskill we would have treat the rating distributions as actual distributions when performing mathematical operations on them (like getting the average of the team). When picking a singular value it is always arbitrary to a degree.
Trueskill assumes that a new player is somewhat average. This is a good assumption in shooters, because they all play more or less the same, but in RTS starting players are *really* bad, because you need to keep up with your economy or you autolose and the ai games don't teach this at all.
We noticed this and coded this interpolation to counteract this problem. Subtracting the deviation is my way to make this interpolation. I think this has the benefit that it utilizes trueskill to determine how fast this interpolation should go away. So it automatically happens slower with games that provide little info about the players' skills.

It could be that 3 * dev is too much. Maybe 2 or 2.5 is better. We should probably look at some player data to see how their winrate for their first games is to get a feeling how big that factor should be.

I already explained this to someone else in the past, so maybe we should add this rationale somewhere in the comments?

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                continue
+            if number_of_players + len(candidate.players) <= team_size * 2:
+                participants.append(candidate)
+                number_of_players += len(candidate.players)
```
I think doing a check for the (3,3,2) edge case where you can't split searches into two teams right here would actually worth considering. I know I've been against that in the past because it feels very inelegant, but given that there is only one edge case (?) for 4v4  and this is the largest supported case I now think it is acceptable.

cleborys replied:
Of course this should happen in a way that doesn't break 5v5, etc. But I think it is ok to only add this optimization for 4v4.

BlackYps replied:
3v3 also has the edge case of (2,2,2). I am not against adding these two, but I am also not convinced what the benefit would be, because we catch these cases later and the matchmaker doesn't seem to be in a strong need for performance optimization.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                number_of_players += len(candidate.players)
+        return participants
+
+    def make_teams(self, searches: List[Search], team_size: int) -> Tuple[Search, Search]:
```
Maybe rename `searches` to `participants` since the former is mostly used for the list of all searches and that's not what's being passed here?

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            raise UnevenTeamsException()
+
+        avg = get_average_rating(searches)
+        team_target_strength = sum(search.cumulated_rating for search in searches) / 2
```
Isn't `team_target_strength` the same as a `avg` above? :D

BlackYps replied:
no, it's the cumulative rating

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+
+    def _run_karmarkar_karp_algorithm(self, searches):
+        self._logger.debug("Running Karmarkar-Karp to partition the teams")
+        # Further reading: https://en.wikipedia.org/wiki/Largest_differencing_method
```
Cool!

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+            raise UnevenTeamsException()
+        return combined_team_a, combined_team_b
+
+    def _run_karmarkar_karp_algorithm(self, searches):
```
Maybe the definition of `Container` could be local to this function? I think that would make a lot of sense since it's pretty specific to this usecase.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        containers_b.append(elem1.content[1])
+        while containers_a or containers_b:
+            for e in containers_a:
+                if len(e.content) == 2:
```
can also make this a method on `Container` so this becomes `if e.is_composed_of_containers` or similar which is a bit less vodoo too ;)

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                    containers_b.append(e.content[1])
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
```
Is it really safe to remove from and append to `containers_a` while iterating over it? A loop that pops from `containers_a` might be safer.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
+            for e in containers_b:
```
You might also consider moving some of this logic into functions on `Container` (but only where you think it makes sense), like
```
next_container = containers_a.pop()
next_container.add_contents_to(team_a, containers_a, containers_b)
``` 
(that might be overkill.)

BlackYps replied:
i think this is indeed overkill. Using your other suggestions already makes the code more straightforward.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+                self._logger.debug("%i players: %s" % (i, searches_by_size[i]))
+        return searches_by_size
+
+    def _find_most_balanced_filler(self, avg, search, searches_dict):
```
I don't think you need to pass the whole `search` or the whole `searches_dict`, isn't this simply finding the player in `searches_dict[1]` whose rating is closest to the target of `team_size * avg - get_cumulative_rating(search)`? Then I would pass that number and `searches_dict[1]` instead.

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        self._logger.debug("used %s as filler", [candidate])
+        return candidate
+
+    def assign_game_quality(self, match: Match) -> GameCandidate:
```
Can you remind me why this isn't the trueskill quality (with some kind of modifications for newbies and time?)
In other words, why aren't we optimizing for trueskill quality?

BlackYps replied:
afaik Trueskill has nothing implemented to account for high rating differences in teams.
[0, 2000] vs. [1000, 1000] would be the same quality like [1000, 1000] vs. [1000, 1000] but we want to prioritze the latter match.

cleborys commented on server/config.py:
```diff
@@ -79,6 +79,18 @@ def __init__(self):
+        self.TIME_BONUS = 0.01
+        self.MAXIMUM_TIME_BONUS = 0.2
+        self.NEWBIE_BONUS = 0.1
+        self.MAXIMUM_NEWBIE_BONUS = 0.4
```
might also call it `NEWBIE_TIME_BONUS` since that's closer to what it does

cleborys commented on server/matchmaker/algorithm/team_matchmaker.py:
```diff
@@ -0,0 +1,314 @@
+        games = SortedList(games, key=lambda game: game.quality)
+
+        matches = []
+        while games:
```
I would suggest the following since it should be faster and I don't think space is a concern
(potentially not valid python names or syntax):
```python
matches = []
used_searches = set()
for game in games:
  if !game.is_disjoint_from(used_searches):
    continue
  matches.append(game.match)
  used_searches += set(game.get_original_searches())

# ...

GameCandidate:
  # ...
  def is_disjoint_from(self, used_searches: Set[Search]):
    for search in self.get_original_searches():
      if search in used_searches:
        return false
    return true
```

cleborys replied:
There's also probably something built in already that can tell you whether `set(game.get_original_searches())` and `used_searches` are disjoint. Might even make `game.get_original_searches` return a set.

cleborys reviewed (commented):
Awesome! I left some  suggestions that I would consider improvements but I think noe of them are necessary if you disagree ;)

BlackYps committed:
Improve readability without touching the logic

BlackYps committed:
Implement refactor suggestions from reviews

BlackYps commented:
This is now ready for review again.
I suggest that we add the per-queue configuration with the database in a separate PR so we don't bloat this one further.

BlackYps committed:
actually use the new fucking code

BlackYps committed:
Sort imports

BlackYps committed:
Tune config values for better results

BlackYps committed:
Add failed matching attempts and newbie status to debug messages

BlackYps committed:
Add sortedcontainers to Pipfile.lock

BlackYps committed:
Make log more readable

BlackYps committed:
Adjust config values

BlackYps committed:
Revert newbie time bonus

cleborys committed:
Feature/#727 refactor matchmaker for 4v4 party algorithm (#749)

* Split matchmaker.algorithm into stable_marriage and bucket_teams

* Make BucketTeamMatchmaker implement Matchmaker

* Make Matchmaker reusable

Move _register_unmatched_searches into matchmaker_queue, add searches as an argument to Matchmaker.find

* Clean up formatting and imports

* Clean up unit test imports, run isort again

* Add some tests for BucketTeamMatchmaker

* Let Matchmaker.find return list of unmatched searches

* Add test for failed matching attempts registered when search is canceled

and minor changes to satisfy static checkers

* Remove unused imports, unused variables

* Fix formatting

* Issue/#801: Register failed matching attempt when queue is too small

* Refactor abstract class with ABC

* Use single extend instead of two appends

* Use hypothesis.assume to filter out uninteresting test data

Co-authored-by: Askaholic <askaholic907@gmail.com>

BlackYps committed:
Fix outdated doc string

BlackYps commented:
The one to be merged is this one: https://github.com/FAForever/server/pull/783
